contracts/src/protocol/UBIPool.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.23;
+
+import { Ownable2StepUpgradeable } from "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import { PubKeyVerification } from "./PubKeyVerification.sol";
+import { IUBIPool } from "../interfaces/IUBIPool.sol";
+
+/// @title UBIPool
+/// @notice Contract for distributing UBI to validators
+/// UBI comes from a percentage of the protocol's emmission during a defined period
+/// This % is minted to the UBIPool contract and can be claimed by validators
+/// Each validator can claim their UBI for a given month
+/// Distributions will be made public monthly by Story for community scrutiny,
+/// and if correct, will be distributed to validators via this contract
+contract UBIPool is IUBIPool, Ownable2StepUpgradeable, ReentrancyGuardUpgradeable, PubKeyVerification {
+    /// @notice The maximum UBI percentage
+    uint32 public immutable MAX_UBI_PERCENTAGE;
+
+    /// @notice The amount of UBI for each validator for each month
+    mapping(uint256 month => mapping(bytes validatorUncmpPubkey => uint256 amount)) public validatorUBIAmounts;
+
+    constructor(uint32 maxUBIPercentage) {
+        MAX_UBI_PERCENTAGE = maxUBIPercentage;
+        _disableInitializers();
+    }
+
+    /// @notice Initializes the contract.
+    /// @param owner The owner of the contract
+    function initialize(address owner) public initializer {
+        require(owner != address(0), "UBIPool: owner cannot be zero address");
+        __Ownable_init(owner);
+    }
+
+    /// @notice Sets the UBI percentage distribution in CL
+    /// @param percentage The percentage of the UBI
+    function setUBIPercentage(uint32 percentage) external onlyOwner {
+        require(percentage < MAX_UBI_PERCENTAGE, "UBIPool: percentage too high");
+        emit UBIPercentageSet(percentage);
+    }
+
+    /// @notice Sets the UBI distribution for a given month
+    /// @param month The month of the distribution (counter, starting from 1, not timestamp or month of year)
+    /// @param totalUBI The total amount of UBI
+    /// @param validatorUncmpPubKeys The validator uncompressed public keys
+    /// @param amounts The amounts of UBI for each validator
+    function setUBIDistribution(
+        uint256 month,
+        uint256 totalUBI,
+        bytes[] calldata validatorUncmpPubKeys,
+        uint256[] calldata amounts
```

Comment:
Since the function can override the amount for a month if it has already been set, it may be better to add a flag, such as `bool override`, to explicitly allow overriding the amount if desired. Alternatively, we could simply revert the transaction to prevent any accidental overrides.

Reply:
We decided to use an internal counter in the contract as distribution id instead of `month`, less chances for input error.
Editing complicates things, there must be careful review of this tx before setting it (maybe timelock).